src/dotnet/App.Maui/Services/ServiceProvider/LazyServiceProviderEngineScope.cs

```diff
@@ -0,0 +1,96 @@
+﻿namespace ActualChat.App.Maui;
+
+public sealed class LazyServiceProviderEngineScope :
+    IServiceScope,
+    IServiceProvider,
+    IServiceScopeFactory,
+    IAsyncDisposable,
+    IDisposable
+{
+    private Task<IServiceProvider> _serviceProviderTask;
+    private readonly bool _isRootScope;
+    private bool _isDisposed;
+    private readonly object _lock = new ();
+    private IServiceProvider? _resolvedServiceProvider;
+
+    public IServiceProvider ServiceProvider => this;
+
+    public LazyServiceProviderEngineScope(Task<IServiceProvider> serviceProviderTask, bool isRootScope)
+    {
+        _serviceProviderTask = serviceProviderTask;
+        _isRootScope = isRootScope;
+    }
+
+    public object? GetService(Type serviceType)
+    {
+        if (serviceType == typeof(IServiceScopeFactory))
+            return this;
+        if (serviceType == typeof(IServiceProvider))
+            return this;
+        var serviceProvider = GetServiceProviderInternal(serviceType);
+        var result = serviceProvider.GetService(serviceType);
+        return result;
+    }
+
+    public void SetupOnResolved(Action<IServiceProvider> setup)
+        => _serviceProviderTask = _serviceProviderTask
+            .ContinueWith(t => {
+                var services = t.Result;
+                setup(services);
+                return services;
+            }, TaskScheduler.Default);
+
+    public IServiceScope CreateScope()
+        => new LazyServiceProviderEngineScope(
+            CreateScopedServiceProvider(),
+            false);
+
+    public void Dispose()
+    {
+        IServiceProvider? serviceProvider;
+        lock (_lock) {
+            if (_isDisposed)
+                return;
+            _isDisposed = true;
+            serviceProvider = _resolvedServiceProvider;
+        }
+        (serviceProvider as IDisposable)?.Dispose();
+    }
+
+    public async ValueTask DisposeAsync()
+    {
+        IServiceProvider? serviceProvider;
+        lock (_lock) {
+            if (_isDisposed)
+                return;
+            _isDisposed = true;
+            serviceProvider = _resolvedServiceProvider;
+        }
+        if (serviceProvider is IAsyncDisposable asyncDisposable)
```

Comment:
maybe ext/util method? I saw similar logic in https://github.com/Actual-Chat/actual-chat/pull/1156/files#diff-08e94eba7ace2273e11f05fb5abb4c029b85eba3c7a52d4abda16cc19b7b88c8R64